File: src/main/java/org/zbus/remoting/Message.java
package org.zbus.remoting;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Serializable;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.StringTokenizer;
import java.util.concurrent.ConcurrentHashMap;
 


public class Message implements Serializable { 
	private static final long serialVersionUID = 4379223525215626137L;

	public static final String HEARTBEAT         = "heartbeat"; //心跳消息
	
	//标准HTTP头部内容
	public static final String HEADER_REMOTE_ADDR= "remote-addr";
	public static final String HEADER_ENCODING   = "content-encoding";
	
	//常见扩展HTTP协议头部
	public static final String HEADER_BROKER     = "broker";
	public static final String HEADER_REMOTE_ID  = "remote-id"; 
	public static final String HEADER_TOPIC      = "topic"; //使用,分隔 
	public static final String HEADER_MQ_REPLY   = "mq-reply";
	public static final String HEADER_MQ         = "mq";
	public static final String HEADER_TOKEN      = "token";
	public static final String HEADER_MSGID      = "msgid";	 //消息ID
	public static final String HEADER_MSGID_SRC  = "msgid-raw"; //原始消息ID
	public static final String HEADER_ACK        = "ack";	
	public static final String HEADER_REPLY_CODE = "reply-code";	 
	public static final String HEADER_WINDOW     = "window";	
	
	
	 
	protected Meta meta = new Meta();
	protected Map<String, String> head = new ConcurrentHashMap<String, String>();
	protected byte[] body; 
	
	
	
	public Message(){
		setBody((byte[])null);
	} 
	
	public static Message copyWithoutBody(Message msg){
		Message res = new Message();
		res.meta = new Meta(msg.meta);
		res.head = new HashMap<String, String>(msg.head);
		res.body = msg.body;
		return res;
	}
	
	public String getMetaString() {
		return meta.toString();
	}
	
	public Meta getMeta(){
		return meta;
	}
	
	public void setMeta(String meta) { 
		this.meta = new Meta(meta);
	}
	
	public void setMeta(Meta meta) { 
		this.meta = meta;
	}
	
	public Map<String, String> getHead() {
		return head;
	} 
	
	public void setHead(Map<String, String> head) {
		this.head = head;
	} 
	
	public String getHead(String key){
		return this.head.get(key);
	}
	
	public void setHead(String key, String value){
		if(value == null) return;
		this.head.put(key, value);
	} 
	
	public String removeHead(String key){
		return this.head.remove(key);
	}
	
	public String getParam(String key){
		return meta.getParam(key); 
	}
	
	public String getHeadOrParam(String key){ 
		String value = getHead(key);
		if(value == null){
			value = getParam(key); 
		} 
		return value;
	}
	
	public String getHeadOrParam(String key, String defaultValue) { 
		String value = getHeadOrParam(key);
		if(value == null){
			value = defaultValue;
		}
		return value;
	}   
	
	public byte[] getBody() {
		return body;
	}
	
	public void setBody(byte[] body) {
		int len = 0;
		if( body != null){
			len = body.length;
		}
		this.setHead("content-length", ""+len);
		this.body = body;
	}
	
	public void setBody(String body){
		setBody(body.getBytes());
	} 
	
	public Message setBody(String format, Object ...args) { 
		this.setBody(String.format(format, args));
		return this;
	} 
	
	public void setJsonBody(String body){
		this.setJsonBody(body.getBytes());
	}
	
	public void setJsonBody(byte[] body){
		this.setHead("content-type", "application/json");
		this.setBody(body);
	}
	
	public String getBodyString() {
		if (this.body == null) return null;
		return new String(this.body);
	}

	public String getBodyString(String encoding) {
		if (this.body == null) return null;
		try {
			return new String(this.body, encoding);
		} catch (UnsupportedEncodingException e) {
			return new String(this.body);
		}
	}
	
	//////////////////////////////////////////////////////////////
	
	public void decodeHeaders(byte[] data, int offset, int size){
		try{
			BufferedReader in = new BufferedReader(new InputStreamReader(new 
				ByteArrayInputStream(data, offset, size)));
	
			String meta = in.readLine();
			if(meta == null) return;
			this.meta = new Meta(meta);
			
			String line = in.readLine();
	        while (line != null && line.trim().length() > 0) {
	            int p = line.indexOf(':');
	            if (p >= 0){ 
	                head.put(line.substring(0, p).trim().toLowerCase(), line.substring(p + 1).trim());
	            } 
	            line = in.readLine();
	        }
		} catch(IOException e){
			e.printStackTrace();
		}
	}
	
	public int estimatedSize(){
		int size = 0;
		size += meta.toString().length();
		
		Iterator<Entry<String, String>> iter = head.entrySet().iterator();
		while(iter.hasNext()){
			Entry<String, String> e = iter.next();
			size += e.getKey().length() + e.getKey().length();
		} 
		
		if(body != null){
			size += body.length;
		}
		return size;
	} 
	
	public String getMqReply() {
		return this.getHeadOrParam(HEADER_MQ_REPLY);
	}
	public Message setMqReply(String value) {
		this.setHead(HEADER_MQ_REPLY, value);
		return this;
	}
	
	public String getEncoding() {
		return this.getHeadOrParam(HEADER_ENCODING);
	}
	public Message setEncoding(String encoding) {
		this.setHead(HEADER_ENCODING, encoding);
		return this;
	}
	
	
	public String getMsgId() {
		return this.getHeadOrParam(HEADER_MSGID);
	}
	public Message setMsgId(String msgId) {
		if(msgId == null) return this;
		this.setHead(HEADER_MSGID, msgId);
		return this;
	}
	
	
	public String getMsgIdSrc() {
		return this.getHeadOrParam(HEADER_MSGID_SRC);
	}
	public Message setMsgIdSrc(String value) {
		if(value == null) return this;
		this.setHead(HEADER_MSGID_SRC, value);
		return this;
	}
	
	public boolean isAck() {
		String ack = this.getHeadOrParam(HEADER_ACK);
		if(ack == null) return true; //默认ack为true
		ack = ack.trim().toLowerCase();
		return ack.equals("1");
	}
	
	public void setAck(boolean ack){
		String value = ack? "1":"0";
		this.setHead(HEADER_ACK, value);
	}
	
	public String getMq(){
		return this.getHeadOrParam(HEADER_MQ);
	}
	 
	public Message setMq(String mq) {
		this.setHead(HEADER_MQ, mq);
		return this;
	} 
	
	public String getToken() {
		return this.getHeadOrParam(HEADER_TOKEN);
	}
	public Message setToken(String token) {
		this.setHead(HEADER_TOKEN, token);
		return this;
	} 
	
	public String getTopic() {
		return getHeadOrParam(HEADER_TOPIC);
	}

	public Message setTopic(String topic) {
		this.setHead(HEADER_TOPIC, topic);
		return this;
	} 
	
	public String getWindow() {
		return getHeadOrParam(HEADER_WINDOW);
	}

	public Message setWindow(int window) {
		this.setHead(HEADER_WINDOW, ""+window);
		return this;
	} 
	

	public String getRemoteId() {
		return getHeadOrParam(HEADER_REMOTE_ID);
	}

	public Message setRemoteId(String value) {
		this.setHead(HEADER_REMOTE_ID, value);
		return this;
	}   


	//////////特殊处理：Command 与 Status 兼容HTTP////////////// 
	public String getCommand() { //特殊处理 
		return meta.command; 
	} 
	public Message setCommand(String command) {
		meta.command = command;
		meta.status = null; //互斥
		return this;
	}   
	public String getStatus() {  
		return meta.status;
	}
	public Message setStatus(String status) { 
		meta.status = status;
		meta.command = null; //互斥
		return this; 
	}
	
	public boolean isStatus200() {
		return "200".equals(this.getStatus());
	}

	public boolean isStatus404() {
		return "404".equals(this.getStatus());
	}

	public boolean isStatus500() {
		return "500".equals(this.getStatus());
	}  
	
	protected String getBodyPrintString() {
		if (this.body == null)
			return null;
		if (this.body.length > 1024) {
			return new String(this.body, 0, 1024) + "...";
		} else {
			return getBodyString();
		}
	}
	
	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder();
		sb.append(meta+"\r\n");
		
		List<String> keys = new ArrayList<String>(head.keySet());
		Collections.sort(keys);
		
		for(String key : keys){ 
			String val = head.get(key);
			sb.append(key+": "+val+"\r\n");
		}
		sb.append("\r\n");
		sb.append(getBodyPrintString());
		return sb.toString();
	} 
	 
}



class Meta implements Serializable{ 
	private static final long serialVersionUID = -8557063231118504061L;
	//HTTP请求头部: 方法(GET/POST)-命令-参数
	String method = "GET"; 
	String command; 
	Map<String,String> params;
	
	//HTTP响应头部: 状态(200)
	String status;
	
	static Set<String> httpMethod = new HashSet<String>();
	static Map<String,String> httpStatus = new HashMap<String, String>();
	
	static{ 
		httpMethod.add("GET");
		httpMethod.add("POST"); 
		httpMethod.add("PUT");
		httpMethod.add("DELETE");
		httpMethod.add("HEAD");
		httpMethod.add("OPTIONS"); 
		
		httpStatus.put("101", "Switching Protocols"); 
		httpStatus.put("200", "OK");
		httpStatus.put("201", "Created");
		httpStatus.put("202", "Accepted");
		httpStatus.put("204", "No Content"); 
		httpStatus.put("206", "Partial Content"); 
		httpStatus.put("301", "Moved Permanently");
		httpStatus.put("304", "Not Modified"); 
		httpStatus.put("400", "Bad Request"); 
		httpStatus.put("401", "Unauthorized"); 
		httpStatus.put("403", "Forbidden");
		httpStatus.put("404", "Not Found"); 
		httpStatus.put("405", "Method Not Allowed"); 
		httpStatus.put("416", "Requested Range Not Satisfiable");
		httpStatus.put("500", "Internal Server Error");
	}
	
	@Override
	public String toString() { 
		if(this.status != null){ //status存在，则认为是响应元数据
			String desc = httpStatus.get(this.status);
			if(desc == null){
				desc = "Unknown Status";
			}
			return String.format("HTTP/1.1 %s %s", status, desc); 
		}
		
		if(this.command != null){//command存在，则认为是请求元数据
			String cmdStr = encodeCommand();
			return String.format("%s /%s HTTP/1.1", method, cmdStr);
		}
		
		return "";
	}
	
	public Meta(){}
	
	public Meta(Meta m){
		this.command = m.command;
		this.method = m.method;
		this.status = m.status;
		if(m.params != null){
			this.params = new HashMap<String, String>(m.params);
		}
	}
	
	public Meta(String meta){
		if("".equals(meta)){
			return;
		}
		StringTokenizer st = new StringTokenizer(meta);
		String method = st.nextToken();
		
		if(!httpMethod.contains(method)){ //不是method字段，理解为Response
			String status = st.nextToken();
			this.status = status;
			return;
		}
		
		this.method = method;  
		String commandString = st.nextToken();
		decodeCommand(commandString);
	} 
	

	private String encodeCommand(){
		StringBuilder sb = new StringBuilder();
		if(this.command != null){
			sb.append(this.command);
		}
		if(this.params == null) return sb.toString();
		
		Iterator<Entry<String, String>> iter = this.params.entrySet().iterator();
		if(iter.hasNext()){
			sb.append("?");
		}
		
		while(iter.hasNext()){
			Entry<String, String> e = iter.next();
			sb.append(e.getKey()+"="+e.getValue());
			if(iter.hasNext()){
				sb.append("&");
			}
		}
		return sb.toString();
	} 
	
	
	private void decodeCommand(String commandString){
		int idx = commandString.indexOf('?');
		if(idx < 0){
			this.command = decodeUrl(commandString);
		} else {
			this.command = commandString.substring(0, idx);
		}
		if(this.command.startsWith("/")){
			this.command = this.command.substring(1);
		}
		if(idx < 0) return;
		
		this.params = new HashMap<String, String>(); 
		String paramString = commandString.substring(idx+1); 
		StringTokenizer st = new StringTokenizer(paramString, "&");
        while (st.hasMoreTokens()) {
            String e = st.nextToken();
            int sep = e.indexOf('=');
            if (sep >= 0) {
                this.params.put(decodeUrl(e.substring(0, sep)).trim(),
                		decodeUrl(e.substring(sep + 1)));
            } else {
                this.params.put(decodeUrl(e).trim(), "");
            }
        } 
	}
	
	private String decodeUrl(String str) {
        String decoded = null;
        try {
            decoded = URLDecoder.decode(str, "UTF8");
        } catch (UnsupportedEncodingException ignored) {
        }
        return decoded;
    }
	
	public String getParam(String key){
		if(params == null) return null;
		return params.get(key);
	}
	
	public String getParam(String key, String defaultValue){
		String value = getParam(key);
		if(value == null){
			value = defaultValue;
		}
		return value;
	}
	
	public void setParam(String key, String value){
		if(value == null) return;
		if(params == null){
			params = new HashMap<String, String>();
		}
		params.put(key, value);
	}
}

File: clients/DotNET/Zbus.Client/Net/Http/Message.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace Zbus.Client.Net.Http
{ 
   public class Message : IId
   {
      public static readonly string REMOTE_ADDR = "remote-addr"; 
      public static readonly string HEARTBEAT = "heartbeat"; 

      public static readonly string CMD = "cmd";
      public static readonly string BROKER = "broker";
      public static readonly string TOPIC = "topic";
      public static readonly string SENDER = "sender";
      public static readonly string RECVER = "recver";
      public static readonly string MQ = "mq";
      public static readonly string ID = "id";
      public static readonly string APPID = "appid";
      public static readonly string TOKEN = "token";
      public static readonly string ACK = "ack";
      public static readonly string ENCODING = "encoding"; //body encoding
      public static readonly string ORIGIN_URL = "origin_url";
      public static readonly string ORIGIN_STATUS = "reply_code";
      public static readonly string ORIGIN_ID = "rawid";

      private Meta meta = new Meta("");
      private IDictionary<string, string> head = new Dictionary<string, string>();
      private byte[] body;

      public string GetHead(string key)
      {
         return GetHead(key, null);
      }

      public string GetHead(string key, string defaultValue)
      {
         string val = null;
         this.head.TryGetValue(key, out val);
         if (val == null)
         {
            val = defaultValue;
         }
         return val;
      }

      public void SetHead(string key, string val)
      {
         this.head[key] = val;
      }

      public void RemoveHead(string key)
      {
         this.head.Remove(key);
      }

      public void SetBody(byte[] body)
      {
         this.body = body;
         int bodyLen = 0;
         if (this.body != null)
         {
            bodyLen = this.body.Length;
         }
         this.SetHead("content-length", string.Format("{0}", bodyLen));
      }

      public void SetJsonBody(string body, System.Text.Encoding encoding)
      {
         this.SetBody(body, encoding);
         this.SetHead("content-type", "application/json");
      }

      public void SetJsonBody(string body)
      {
         SetJsonBody(body, System.Text.Encoding.Default);
      }

      public void SetBody(string body, System.Text.Encoding encoding)
      {
         Encoding = encoding.WebName;
         SetBody(encoding.GetBytes(body));
      }

      public void SetBody(string body)
      {
         SetBody(body, System.Text.Encoding.Default);
      }

      public void SetBody(string format, params object[] args)
      {
         SetBody(string.Format(format, args));
      }
       
      #region AUX_GET_SET

      public String Mq
      {
         get { return GetHead(MQ); }
         set { SetHead(MQ, value); }
      }

      public String Cmd
      {
         get { return GetHead(CMD); }
         set { SetHead(CMD, value); }
      } 

      public String Id
      {
         get { return GetHead(ID); }
         set { SetHead(ID, value); }
      }

      public String AppId
      {
         get { return GetHead(APPID); }
         set {  SetHead(APPID, value); }
      }

      public String Token
      {
         get { return GetHead(TOKEN); }
         set { SetHead(TOKEN, value); }
      }

      public String Sender
      {
         get { return GetHead(SENDER); }
         set { SetHead(SENDER, value); }
      }

      public String Recver
      {
         get { return GetHead(RECVER); }
         set {  SetHead(RECVER, value);  }
      }
      public String Topic
      {
         get { return GetHead(TOPIC); }
         set {  SetHead(TOPIC, value); }
      }

      public String Encoding
      {
         get { return GetHead(ENCODING); }
         set { SetHead(ENCODING, value); }
      }

      public String OriginUrl
      {
         get { return GetHead(ORIGIN_URL); }
         set { SetHead(ORIGIN_URL, value); }
      }

      public String OriginStatus
      {
         get { return GetHead(ORIGIN_STATUS); }
         set { SetHead(ORIGIN_STATUS, value); }
      }

      public String OriginId
      {
         get { return GetHead(ORIGIN_ID); }
         set { SetHead(ORIGIN_ID, value); }
      }

      public bool Ack
      {
         get
         {
            string ack = GetHead(ACK);
            return (ack == null) || "1".Equals(ack); //default to true 
         }
         set { SetHead(ACK, value ? "1" : "0"); }
      }

      public string Url
      {
         get { return this.meta.Url; }
         set
         {
            this.meta.Url = value;
            this.meta.Status = null;
         }
      }

      public string RequestPath
      {
         get { return this.meta.RequestPath; }
         set { this.meta.RequestPath = value; }
      }

      public IDictionary<string, string> RequestParams
      {
         get { return this.meta.RequestParams; }
         set { this.meta.RequestParams = value; }
      }

      public string Status
      {
         get { return this.meta.Status; }
         set { this.meta.Status = value; }
      } 

      public byte[] Body
      {
         get { return body; }
         set { SetBody(value); }
      }

      public string BodyString
      {
         get { return GetBody(); }
         set { SetBody(value); }
      }

      public string BodyJson
      {
         get { return GetBody(); }
         set { SetJsonBody(value); }
      }


      #endregion

      public void Encode(IoBuffer buf)
      {
         buf.Put("{0}\r\n", this.meta);
         foreach (KeyValuePair<string, string> e in this.head)
         {
            buf.Put("{0}: {1}\r\n", e.Key, e.Value);
         }
         string lenKey = "content-length";
         if (!this.head.ContainsKey(lenKey))
         {
            int bodyLen = this.body == null ? 0 : this.body.Length;
            buf.Put("{0}: {1}\r\n", lenKey, bodyLen);
         }

         buf.Put("\r\n");

         if (this.body != null)
         {
            buf.Put(this.body);
         }
      }

      private static int FindHeaderEnd(IoBuffer buf)
      {
         int i = buf.Position;
         byte[] data = buf.Data;
         while (i + 3 < buf.Limit)
         {
            if (data[i] == '\r' && data[i + 1] == '\n' && data[i + 2] == '\r' && data[i + 3] == '\n')
            {
               return i + 3;
            }
            i++;
         }
         return -1;

      }

      public static Message DecodeHeader(string header)
      {
         Message msg = new Message();
         string[] lines = Regex.Split(header, "\r\n");
         msg.meta = new Meta(lines[0]);
         for (int i = 1; i < lines.Length; i++)
         {
            string line = lines[i];
            int idx = line.IndexOf(':');
            if (idx < 0) continue; //ignore
            string key = line.Substring(0, idx).Trim().ToLower(); //key to lower case
            string val = line.Substring(idx + 1).Trim();
            msg.SetHead(key, val);
         }

         return msg;
      }

      public static Message Decode(IoBuffer buf)
      {
         int idx = FindHeaderEnd(buf);
         int headLen = idx - buf.Position + 1;
         if (idx < 0) return null;

         string header = System.Text.Encoding.Default.GetString(buf.Data, buf.Position, headLen);

         Message msg = DecodeHeader(header);
         string bodyLenString = msg.GetHead("content-length");
         if (bodyLenString == null)
         {
            buf.Drain(headLen);
            return msg;
         }
         int bodyLen = int.Parse(bodyLenString);
         if (buf.Remaining() < headLen + bodyLen)
         {
            return null;
         }
         buf.Drain(headLen);
         byte[] body = buf.Get(bodyLen);
         msg.SetBody(body);
         return msg;
      }

      public override string ToString()
      {
         IoBuffer buf = new IoBuffer();
         Encode(buf);
         return buf.ToString();
      }

      public string GetBody()
      {
         if (body == null) return null;
         return System.Text.Encoding.Default.GetString(body);
      }

      public string GetBody(System.Text.Encoding encoding)
      {
         if (body == null) return null;
         return encoding.GetString(body);
      }
   }

   class Meta
   {
      static readonly HashSet<string> httpMethod = new HashSet<string>();
      static IDictionary<string, string> httpStatus = new Dictionary<string, string>();
      static Meta()
      {
         httpMethod.Add("GET");
         httpMethod.Add("POST");
         httpMethod.Add("HEAD");
         httpMethod.Add("PUT");
         httpMethod.Add("DELETE");
         httpMethod.Add("OPTIONS");

         httpStatus.Add("200", "OK");
         httpStatus.Add("201", "Created");
         httpStatus.Add("202", "Accepted");
         httpStatus.Add("204", "No Content");
         httpStatus.Add("206", "Partial Content");
         httpStatus.Add("301", "Moved Permanently");
         httpStatus.Add("304", "Not Modified");
         httpStatus.Add("400", "Bad Request");
         httpStatus.Add("401", "Unauthorized");
         httpStatus.Add("403", "Forbidden");
         httpStatus.Add("404", "Not Found");
         httpStatus.Add("405", "Method Not Allowed");
         httpStatus.Add("416", "Requested Range Not Satisfiable");
         httpStatus.Add("500", "Internal Server Error");
      }

      public string Method { get; set; }
      public string Status { get; set; }
      public string Url
      {
         get
         {
            return _url;
         }
         set
         {
            _url = value;
            Status = null;
         }
      }
      public string RequestPath
      {
         get
         {
            return _path;
         }
         set
         {
            this._path = value;
            CalcUrl();
         }
      }

      public IDictionary<string, string> RequestParams
      {
         get
         {
            return _requestParams;
         }
         set
         {
            this._requestParams = value;
            CalcUrl();
         }
      }

      private string _url;
      private string _path;
      private IDictionary<string, string> _requestParams;
      public string GetParam(string key)
      {
         return GetParam(key, null);
      }

      public string GetParam(string key, string defaultValue)
      {
         if (this._requestParams == null)
            return defaultValue;
         string value;
         this._requestParams.TryGetValue(key, out value);
         if (value == null)
            value = defaultValue;
         return value;
      }

      public void SetParam(string key, string val)
      {
         if (this._requestParams == null)
         {
            this._requestParams = new Dictionary<string, string>();
         }
         this._requestParams[key] = val;
         CalcUrl();
      }



      public Meta(string meta)
      {
         this._url = "/";
         this.Method = "GET";
         if (meta == null || meta.Trim().Equals("")) return;
         string[] blocks = meta.Split(' ');
         string method = blocks[0];
         if (!httpMethod.Contains(method))
         {
            this.Status = blocks[1];
            return;
         }
         this.Method = method;
         this.DecodeUrl(blocks[1]);
      }

      public override string ToString()
      {
         if (this.Status != null)
         {
            string desc = null;
            httpStatus.TryGetValue(this.Status, out desc);
            if (desc == null)
            {
               desc = "Unknown Status";
            }
            return string.Format("HTTP/1.1 {0} {1}", this.Status, desc);
         }
         else
         {
            return string.Format("{0} {1} HTTP/1.1", this.Method, this.Url);
         }
      }

      private void CalcUrl()
      {
         StringBuilder sb = new StringBuilder();
         if (_path != null)
         {
            sb.Append(_path);
         }
         if (_requestParams != null)
         {
            sb.Append("?");
            var iter = _requestParams.GetEnumerator();
            bool first = true;
            while (iter.MoveNext())
            {
               if (!first)
               {
                  sb.Append("&&");
               }
               else
               {
                  first = false;
               }

               sb.Append(iter.Current.Key + "=" + iter.Current.Value);
            }
         }
         this.Url = sb.ToString();

      }

      public void DecodeUrl(string url)
      {
         this._url = url;

         int idx = url.IndexOf('?');
         if (idx < 0)
         {
            this._path = url;
         }
         else
         {
            this._path = url.Substring(0, idx);
         }
         if (this._path[0] == '/')
         {
            this._path = this._path.Substring(1);
         }

         if (idx < 0) return;
         string args = url.Substring(idx + 1);
         this._requestParams = new Dictionary<string, string>();
         string[] blocks = args.Split('&');
         foreach (string kv in blocks)
         {
            string[] kvb = kv.Split("=".ToCharArray(), 2);
            this._requestParams[kvb[0].Trim()] = kvb[1].Trim();
         }
      }

   }
}
